ledger/participant-integration-api/src/main/scala/platform/indexer/MeteringAggregator.scala

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.indexer
+
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.store.appendonlydao.DbDispatcher
+import com.daml.platform.store.backend.{MeteringStorageWriteBackend, ParameterStorageBackend}
+import com.daml.scalautil.Statement.discard
+import java.time.{OffsetDateTime, ZoneOffset}
+import java.util.{Timer, TimerTask}
+import scala.concurrent.duration._
+
+object MeteringAggregator {
+
+  class Owner(
+      meteringStore: MeteringStorageWriteBackend,
+      parameterStore: ParameterStorageBackend,
+      period: FiniteDuration = 10.minutes,
+  ) {
+
+    private[platform] def apply(
+        dispatcher: DbDispatcher
+    )(implicit loggingContext: LoggingContext): ResourceOwner[Unit] = {
+      val aggregator = new MeteringAggregator(meteringStore, parameterStore)
+      ResourceOwner.forTimer(() => new Timer()).map { timer =>
+        timer.schedule(
+          new TimerTask {
+            override def run(): Unit = aggregator.run(dispatcher)
```

Comment:
You can just simply add DbDispatcher to the class MeteringAggregator, since dbDispatcher is already available as you instantiate it (or doing the functional approach, having metering aggregator just an object with a function which does something...my point: having them on different levels is not needed)

Reply:
I would be slightly concerned about dispatcher is going down when metering aggregator is doing something... making it somehow cancellable/releaseable (eg a ResourceOwner) would help probably (not sure it is feasible, but at least we could do something like letting it finish before releasing the DbDispatcher)

Reply:
There is always the chance of a database connection being lost during processing.  The aggregator uses a DB transaction to ensure the operation is never part completed.  On restart (HA or not) the aggregator will re-run and process and outstanding aggregations.